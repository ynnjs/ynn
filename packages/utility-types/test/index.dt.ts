/******************************************************************
 * Copyright (C) 2020 LvChengbin
 * 
 * File: test/index.spec.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 12/27/2020
 * Description: 
 ******************************************************************/

/* eslint @typescript-eslint/no-unused-vars: 'off' */

import {
    VariadicFunction,
    VariadicClass,
    ParametersShift,
    Diff,
    PartialKeys,
    PartialExcludesKeys
} from '../src';

function testType<T>(): T {
    return undefined as any;
}

// @dts-jest:group VariadicFunction<P = any[], R = any>
{

    interface O {
        f1?: VariadicFunction;
        f2?: VariadicFunction<[string, number], number>;
    }

    const o: O = {};

    // @dts-jest:pass:snap default arguments
    testType<VariadicFunction>();

    // @dts-jest:pass:snap specific arguments
    testType<VariadicFunction<[string, number], number>>();

    // @dts-jest:pass:snap default arguments
    o.f1 = () => {};

    // @dts-jest:pass:snap specific arguments
    o.f2 = ( a: string, b: number ) => b;
}

// @dts-jest:group VariadicClass<P = any[], T = any>
{

    interface X {
        name: string;
    }

    interface O {
        c1?: VariadicClass;
        c2?: VariadicClass<[string, number], X>
    }

    const o: O = {};

    // @dts-jest:pass:snap variadic arguments and return type
    testType<VariadicClass>();

    // @dts-jest:pass:snap specific arguments
    testType<VariadicClass<[string, number], X>>();

    // @dts-jest:pass:snap default arguments
    o.c1 = class {}

    // @dts-jest:pass:snap specific arguments
    o.c2 = class {
        constructor( public name: string, public age: number ) {}
    }

}

// @dts-jest:group ParametersTail<T extends VariadicFunction, M extends any[] = [ any ]>
{
    interface Fn {
        ( a: string, b: number, c: boolean ): void;
    }

    // @dts-jest:pass:snap should remove the first item in parameters
    const x: ParametersShift<Fn> = [ 123, true ];

    interface Fnn {
        (): void;
    }

    // @dts-jest:pass:snap should return empty parameter list
    const y: ParametersShift<Fnn> = [];
}

// @dts-jest:group Diff<T, U>
{
    // @dts-jest:pass
    const x: Diff<'a' | 'b' | 'c', 'a' | 'b'> = 'c';

    // @dts-jest:fail
    const y: Diff<'a' | 'b' | 'c', 'a' | 'b'> = 'b';
}

// @dts-jest:group PartialKeys<T, M extends keyof T>
{
    interface O {
        name: string;
        age: number;
        sex: number;
    }

    // @dts-jest:fail
    const x: PartialKeys<O, 'name' | 'age'> = {}

    // @dts-jest:fail
    const y: PartialKeys<O, 'name' | 'age'> = { name : 'x' }

    // @dts-jest:pass:snap should make all specific keys optional
    const z: PartialKeys<O, 'name' | 'age'> = { sex : 1 }
}

// @dts-jest:group PartialExcludesKeys<T, M extends keyof T>
{
    interface O {
        name: string;
        age: number;
        sex: number;
    }

    // @dts-jest:fail
    const x: PartialExcludesKeys<O, 'name' | 'age'> = {}

    // @dts-jest:fail
    const y: PartialExcludesKeys<O, 'name' | 'age'> = { name : 'x' }

    // @dts-jest:pass:snap should make all other keys optional
    const z: PartialExcludesKeys<O, 'name' | 'age'> = { name : 'x', age : 1 }
}
