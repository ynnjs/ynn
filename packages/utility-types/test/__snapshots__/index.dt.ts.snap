// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParametersTail<T extends VariadicFunction, M extends any[] = [ any ]> should remove the first item in parameters (type) should match snapshot 1`] = `"any"`;

exports[`ParametersTail<T extends VariadicFunction, M extends any[] = [ any ]> should return empty parameter list (type) should match snapshot 1`] = `"any"`;

exports[`ParametersTail<T extends VariadicFunction, M extends any[] = [ any ]> the return type should be an empty array (type) should match snapshot 1`] = `
"Type '[unknown]' is not assignable to type '[]'.
  Source has 1 element(s) but target allows only 0."
`;

exports[`PartialExcludesKeys<T, M extends keyof T> should make all other keys optional (type) should match snapshot 1`] = `"any"`;

exports[`PartialKeys<T, M extends keyof T> should make all specific keys optional (type) should match snapshot 1`] = `"any"`;

exports[`RequiredExcludesKeys<T, U extends keyof T> const b: RequiredExcludesKeys<O, 'name' | 'age'> = { name : 'x', age : 1 } (type) should match snapshot 1`] = `
"Type '{ name: string; age: number; }' is not assignable to type 'RequiredExcludesKeys<O, \\"name\\" | \\"age\\">'.
  Property 'sex' is missing in type '{ name: string; age: number; }' but required in type 'Required<Pick<O, \\"sex\\">>'."
`;

exports[`VariadicClass<P = any[], T = any> default arguments (type) should match snapshot 1`] = `"typeof c1"`;

exports[`VariadicClass<P = any[], T = any> specific arguments (type) should match snapshot 1`] = `"VariadicClass<[string, number], X>"`;

exports[`VariadicClass<P = any[], T = any> specific arguments (type) should match snapshot 2`] = `"typeof c2"`;

exports[`VariadicClass<P = any[], T = any> the first arguments should be a type extends from any[] (type) should match snapshot 1`] = `"Type 'string' does not satisfy the constraint 'any[]'."`;

exports[`VariadicClass<P = any[], T = any> variadic arguments and return type (type) should match snapshot 1`] = `"VariadicClass<any[], any>"`;

exports[`VariadicFunction<P = any[], R = any> default arguments (type) should match snapshot 1`] = `"VariadicFunction<unknown[], unknown>"`;

exports[`VariadicFunction<P = any[], R = any> default arguments (type) should match snapshot 2`] = `"() => void"`;

exports[`VariadicFunction<P = any[], R = any> specific arguments (type) should match snapshot 1`] = `"VariadicFunction<[string, number], number>"`;

exports[`VariadicFunction<P = any[], R = any> specific arguments (type) should match snapshot 2`] = `"(a: string, b: number) => number"`;

exports[`VariadicFunction<P = any[], R = any> the first argument should be a type extends from any[] (type) should match snapshot 1`] = `"Type 'string' does not satisfy the constraint 'unknown[]'."`;

exports[`WritableExcludesKeys<T, K extends keyof T> a.name = 'y' (type) should match snapshot 1`] = `"Cannot assign to 'name' because it is a read-only property."`;

exports[`WritableKeys<T, K extends keyof T> a.age = 2 (type) should match snapshot 1`] = `"Cannot assign to 'age' because it is a read-only property."`;
